processing/src/main/java/org/apache/druid/query/DefaultQueryMetrics.java

```diff
@@ -23,51 +23,64 @@
 import com.fasterxml.jackson.databind.ObjectMapper;
 import com.google.common.collect.ImmutableMap;
 import org.apache.druid.collections.bitmap.BitmapFactory;
+import org.apache.druid.java.util.common.ISE;
 import org.apache.druid.java.util.common.StringUtils;
 import org.apache.druid.java.util.emitter.service.ServiceEmitter;
 import org.apache.druid.java.util.emitter.service.ServiceMetricEvent;
 import org.apache.druid.query.filter.Filter;
 import org.joda.time.Interval;
 
+import javax.annotation.concurrent.GuardedBy;
+import java.util.Arrays;
 import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
 import java.util.concurrent.TimeUnit;
 
-/**
- * DefaultQueryMetrics is unsafe for use from multiple threads. It fails with RuntimeException on access not from the
- * thread where it was constructed. To "transfer" DefaultQueryMetrics from one thread to another {@link #ownerThread}
- * field should be updated.
- */
 public class DefaultQueryMetrics<QueryType extends Query<?>> implements QueryMetrics<QueryType>
 {
   protected final ObjectMapper jsonMapper;
-  protected final ServiceMetricEvent.Builder builder = new ServiceMetricEvent.Builder();
-  protected final Map<String, Number> metrics = new HashMap<>();
+  protected final Object lock = new Object();
+  @GuardedBy("lock") private final Map<String, String> singleValueDims = new HashMap<>();
+  @GuardedBy("lock") private final Map<String, String[]> multiValueDims = new HashMap<>();
+  @GuardedBy("lock") private final Map<String, Number> metrics = new HashMap<>();
+  @GuardedBy("lock") private Thread ownerThread;
```

Comment:
Should remain `protected` for subclasses

Reply:
subclasses are supposed to use setDimension/setDimensions/reportMetric methods (with proper synchronization and check) instead of directly manipulating on these fields

Reply:
I mean specifically the `ownerThread` field.